crates/worker/src/partition_processor_manager/mod.rs

```diff
@@ -538,24 +567,48 @@ impl PartitionProcessorManager {
     }
 
     fn on_command(&mut self, command: ProcessorsManagerCommand) {
-        use ProcessorsManagerCommand::*;
         match command {
-            CreateSnapshot(partition_id, sender) => {
-                if let Some(processor_state) = self.processor_states.get(&partition_id) {
-                    processor_state.create_snapshot(sender);
-                } else {
-                    let _ = sender.send(Err(anyhow::anyhow!(
-                        "Partition processor '{}' not found",
-                        partition_id
-                    )));
-                }
+            ProcessorsManagerCommand::CreateSnapshot(partition_id, sender) => {
+                self.handle_create_snapshot(partition_id, sender);
             }
-            GetState(sender) => {
+            ProcessorsManagerCommand::GetState(sender) => {
                 let _ = sender.send(self.get_state());
             }
         }
     }
 
+    fn handle_create_snapshot(
+        &mut self,
+        partition_id: PartitionId,
+        sender: oneshot::Sender<SnapshotResult>,
+    ) {
+        let processor_state = match self.processor_states.get(&partition_id) {
+            Some(state) => state,
+            None => {
+                let _ = sender.send(Err(SnapshotError::PartitionNotFound(partition_id)));
+                return;
+            }
+        };
+
+        if !processor_state.can_create_snapshot() {
+            let _ = sender.send(Err(SnapshotError::InvalidState(partition_id)));
+            return;
+        }
+
+        let archived_lsn_sender = match self.archived_lsn_channels.get(&partition_id).cloned() {
+            Some(watch) => watch.0.clone(),
+            None => {
+                let _ = sender.send(Err(SnapshotError::Internal(
+                    partition_id,
+                    "No archived LSNs channel found!".to_string(),
+                )));
+                return;
+            }
+        };
```

Comment:
Not sure whether I fully understand why we still need the watch channels to communicate the archived lsn back to the `PartitionProcessorManager`. Seems easier to return it directly from the task.